### PR TITLE
Added a contribution issue

### PR DESCRIPTION
**Pull request by akoszegi**
Added an issue template for users to open when they have an example they want to contribute. 

@m3ller Please review and make any changes you see fit.

### CODE REVIEW

File: .github/CODEOWNERS.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# For details on CODEOWNERS file, see
```
Comment (JoelPasvolsky) [inline .github/CODEOWNERS.txt]: Should CODEOWNERS have a ".txt" extension?

File: .github/CODEOWNERS.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+# https://help.github.com/en/articles/about-code-owners
+
+# Default owner of this repo
```
Comment (m3ller) [inline .github/CODEOWNERS.txt]: I don't believe we are adding the original author as a code owner. According to the code owner link that you have in this doc, the code owner needs write permissions to this repo. I would be surprised if we are giving them write permissions to the repo.

What write permission would entail: https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/repository-permission-levels-for-an-organization#repository-access-for-each-permission-level

File: .github/ISSUE_TEMPLATE/contribute.txt [hunk 3]
```diff
@@ -0,0 +1,24 @@
+---
+name: Contribute
+about: Let us know you have an example you'd like to contribute
+
```
Comment (JoelPasvolsky) [inline .github/ISSUE_TEMPLATE/contribute.txt]: These two lines are unclear to me (as full line -- but see below):
* Should they leave "Contribute" as the name of this file or do they replace with a name for their example? 
* How does "Let us ..." relate to "About"? I'd have expected something like "Tell us about the project you wish to contribute." 
Also, "name" and "about" should be "Name" and "About" (again, see below)

Maybe the post-colon text pops up in GitHub by itself (and "name" and "about" are just keywords)? If so, then just needs a period after "like to contribute"

File: .github/ISSUE_TEMPLATE/contribute.txt [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+---
+
+**Description**
+A clear and concise description of the concepts in your project and the problem your code solves. Please include any references you used.
```
Comment (m3ller) [inline .github/ISSUE_TEMPLATE/contribute.txt]: Perhaps break the line at column 80?

File: .github/ISSUE_TEMPLATE/contribute.txt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+---
+
+**Description**
+A clear and concise description of the concepts in your project and the problem your code solves. Please include any references you used. 
+
+**How to run the example**
+A short description of how to run your code, including an example. 
+
+**Link to your repo**
+Include a link to the repo you want to contribute. We will fork it. If we want to make any changes to the code we will create a pull request for you to approve.
```
Comment (JoelPasvolsky) [inline .github/ISSUE_TEMPLATE/contribute.txt]: Do we want to always be making desired code changes ourselves, without the option of requesting that they modify the code per our preferences? Seems like we might be getting a lot of work from this.

File: .github/ISSUE_TEMPLATE/contribute.txt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+**Environment**
+ - Python version: [e.g., 3.7.0]
+ - Ocean version: [e.g., 2.0]
+
+**Additional Context**
```
Comment (JoelPasvolsky) [inline .github/ISSUE_TEMPLATE/contribute.txt]: "Additional information" (non-cap to be consistent with previous one) or "Context"?

File: .github/ISSUE_TEMPLATE/contribute.txt (hunk 5 above)
Comment (m3ller) [inline .github/ISSUE_TEMPLATE/contribute.txt]: Perhaps be more explicit that the repo they are contributing is going into `dwave-examples`?
"Include a link... to contribute to `dwave-examples`"

Comment (m3ller) [inline .github/ISSUE_TEMPLATE/contribute.txt]: "If we want to make any changes to the code we will create a pull request for you to approve." Seems like something that we would have to be held to in the future...probably not what we want.

Perhaps simply suggest that we may make a pull request to their repo? "We may make a pull request to your repo if we would like your approval for certain code changes." Not sure about wording...will like to hear feedback from other reviewers.

**Comment by m3ller** [comment]
Closes #2

File: .github/ISSUE_TEMPLATE/contribute.txt [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+Add any other information.
+
+Before submitting, please take a look at our How To Contribute guide:
+https://docs.ocean.dwavesys.com/en/latest/contributing.html
```
Comment (m3ller) [inline .github/ISSUE_TEMPLATE/contribute.txt]: Perhaps also tell them to follow the outline in this `template` repo?

File: .github/CODEOWNERS.txt (hunk 1 above)
Comment (randomir) [reply .github/CODEOWNERS.txt]: I believe it should be `md`, for markdown, but maybe GH supports `txt` as well. :man_shrugging:

File: .github/ISSUE_TEMPLATE/contribute.txt (hunk 3 above)
Comment (randomir) [reply .github/ISSUE_TEMPLATE/contribute.txt]: This is just GH's header in the markdown doc.

File: .github/CODEOWNERS.txt (hunk 2 above)
Comment (m3ller) [reply .github/CODEOWNERS.txt]: My comment was based on thinking that we would be forking from their repo:
- in the case of forking, we could simply rebase from their original repo without giving them write-access to our fork
- as this would be an external contributor, I am not sure if we want to give them free rein (write-access) on what is written in the `dwave-examples` repo without someone internally having reviewed it first. Even if they are the original author

In the context of a repo transfer:
- I'm uncertain if we could do repo transfers as "To transfer a repository that you own to an organization, you must have permission to create a repository in the target organization." Perhaps we could try with a dummy account that is not a member of the `dwave-examples` organization?  https://help.github.com/en/github/administering-a-repository/transferring-a-repository#about-repository-transfers

File: .github/CODEOWNERS.txt (hunk 1 above)
Comment (akoszegi) [reply .github/CODEOWNERS.txt]: oops, I missed that. Thanks for catching it!

File: .github/CODEOWNERS.txt (hunk 2 above)
Comment (randomir) [reply .github/CODEOWNERS.txt]: That's a good catch, we should verify it's possible to transfer a repo to dwave-examples without giving them write access to org.

Also, I didn't point this out, but obviously, to give someone a write access to a repo under our org, we should make sure that user is trustworthy.

File: .github/ISSUE_TEMPLATE/contribute.txt (hunk 5 above)
Comment (akoszegi) [reply .github/ISSUE_TEMPLATE/contribute.txt]: @m3ller @randomir what do you think?

Comment (akoszegi) [reply .github/ISSUE_TEMPLATE/contribute.txt]: I'll add both options to the template

File: .github/CODEOWNERS.txt (hunk 2 above)
Comment (akoszegi) [reply .github/CODEOWNERS.txt]: I added both options. Let me know if you'd like help testing the ownership transfer option.

File: .github/ISSUE_TEMPLATE/contribute.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+There are two options for how to contribute your example:
+1. Forking  
+    D-Wave will fork your repo. We may make a pull request to your repo if we would 
+	like your approval for certain code changes.
+2. Transferring repo ownership
+    You may choose to transfer ownership of your repo to D-Wave. Here's more 
+	information on this option: 
+	
+	https://help.github.com/en/github/administering-a-repository/transferring-a-repository
```
Comment (m3ller) [inline .github/ISSUE_TEMPLATE/contribute.md]: In the collection of examples, will this repo show up as a "D-Wave" example or an external contribution? I'm not sure how that differentiation occurs in the metadata-grabbing code. @randomir

File: .github/CODEOWNERS.txt (hunk 2 above)
Comment (m3ller) [reply .github/CODEOWNERS.txt]: Yes, I'd like help testing out the ownership transfer option, @akoszegi. Since you're not a member of the `dwave-examples` organization, could you please try transferring a dummy repo to the organization?

Comment (m3ller) [reply .github/CODEOWNERS.txt]: If it works, I would be happy allowing both options (as you have written @akoszegi). If problems occur, we can always update this boilerplate.

File: .github/ISSUE_TEMPLATE/contribute.md (hunk 8 above)
Comment (randomir) [reply .github/ISSUE_TEMPLATE/contribute.md]: Good point. It'll show up as owned by D-Wave. Right now, I'm not sure we can differentiate it on the examples page, but I'll double-check when you test out the transfer.

File: .github/CODEOWNERS.txt (hunk 2 above)
Comment (akoszegi) [reply .github/CODEOWNERS.txt]: I got this error when I tried. Is there a way around that?
image

Comment (m3ller) [reply .github/CODEOWNERS.txt]: Hi @akoszegi and @randomir . I don't see a way around it. The documentation (see above "about repository transfers" link) was specific that one of the prerequisites for a repo transfer was that the person needs permission to create a repo in the organization. 

Since "outside collaborators can never create repositories" and only members of the organization have the chance to create repos, I don't think the external repo transfer idea is possible. (see picture in step 5 of https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/restricting-repository-creation-in-your-organization)

**tl;dr**: external contributor would need to become a member. Owner would need to give members the power to create repos. Don't think this is a scale-able solution.

note: we could always give the original author write-permissions to our fork of their repo. I don't see any technical github issues there, only potential branding issues.

Comment (m3ller) [reply .github/CODEOWNERS.txt]: @akoszegi: thanks for testing and confirming the issue

Comment (randomir) [reply .github/CODEOWNERS.txt]: Thanks for testing this, @akoszegi and @m3ller. Let's keep it simple for now and talk about forking only.

Personally, I don't see a problem in inviting external users to `dwave-examples` -- it's just an account for D-Wave/Ocean-relevant examples, not our main GitHub account. But doing that would surely require a broader discussion.

Comment (m3ller) [reply .github/CODEOWNERS.txt]: We have to pay per member for `dwave-examples`, unfortunately. Hence, my not scale-able comment.

Comment (akoszegi) [reply .github/CODEOWNERS.txt]: I changed it to just include forking.

**Comment by m3ller** [comment]
Looks good. I'm not sure where we are at with write-permissions to the fork, but we can do that by ear. I think this boilerplate will improve the repo, so I will just merge.
